### PR TITLE
Add :db.fn/retractAttribute and :db.fn/retractEntity. Fixes #46.

### PR DESCRIPTION
**Pull request by ncalexan**
@rnewman this is ready for your review.

### CODE REVIEW

File: test/datomish/db_test.cljc [hunk 1]
```diff
@@ -81,541 +81,445 @@
...
+                                    [:db/add (d/id-literal :db.part/user -2) :spouse "Dana"]])))))))
+
+(deftest-db test-valueType-keyword conn
+  (let [tx0 (:tx (<? (d/<transact! conn [{:db/id        (d/id-literal :db.part/user -1)
+                                          :db/ident     :test/kw
+                                          :db/unique    :db.unique/identity
+                                          :db/valueType :db.type/keyword}
+                                         {:db/id :db.part/db :db.install/attribute (d/id-literal :db.part/user -1)}])))]
+
+    (let [report (<? (d/<transact! conn [[:db/add (d/id-literal :db.part/user -1) :test/kw :test/kw1]]))
```
Comment (rnewman) [inline test/datomish/db_test.cljc]: `let` is sequential in Clojure, so you can actually mash the nested ones together (throughout).

File: src/datomish/db.cljc [hunk 2]
```diff
@@ -172,15 +172,63 @@
...
+        (cons
+          (apply str
+                 "INSERT INTO tx_lookup (e0, a0, v0, tx0, added0, value_type_tag0, sv, svalue_type_tag)
+                SELECT e, a, v, ?, 0, value_type_tag, v, value_type_tag
+                FROM datoms
+                WHERE "
+                 (repeater (count chunk)))
+          (cons
+            tx
+            (mapcat (fn [[_ e a]]
```
Comment (rnewman) [inline src/datomish/db.cljc]: So I guess these aren't `eas`, then — they're `[source e a tx]`?

File: src/datomish/db.cljc [hunk 3]
```diff
@@ -172,15 +172,63 @@
...
+                      [e a])
+                    chunk))))
+      (partition-all (quot (dec max-sql-vars) 2) eas))))
+
+;; TODO: make this not do the tx_lookup.  We could achieve this by having additional special values
+;; of added0, or by separating the tx_lookup table into before and after tables.
+(defn- retractEntities->queries [es tx]
+  (let [where-part
+        "e = ? OR (v = ? AND value_type_tag = ?)"
```
Comment (rnewman) [inline src/datomish/db.cljc]: A tiny comment here: that we're retracting all tuples for which the provided entity is either the entity or value.

And you can save a var by concatenating ref-tag here directly into the string — it's known to be an integer.

File: src/datomish/db.cljc [hunk 4]
```diff
@@ -172,15 +172,63 @@
...
+                SELECT e, a, v, ?, 0, value_type_tag, v, value_type_tag
+                FROM datoms
+                WHERE "
+                 (repeater (count chunk)))
+          (cons
+            tx
+            (mapcat (fn [[_ e]]
+                      [e e ref-tag])
+                    chunk))))
+      (partition-all (quot (dec max-sql-vars) 3) es))))
```
Comment (rnewman) [inline src/datomish/db.cljc]: … so this can be `2`, and we can happily skip the `ref-tag` above.

File: src/datomish/db.cljc [hunk 5]
```diff
@@ -431,16 +479,27 @@
...
         operations (group-by first entities)]
 
-    (when-not (clojure.set/subset? (keys operations) #{:db/retract :db/add})
-      (raise (str "Unknown operations " (keys operations))
-             {:error :transact/syntax, :operations (dissoc operations :db/retract :db/add)}))
+    ;; Belt and braces.  At this point, we should have already errored out if op is not known.
+    (let [known #{:db/retract :db/add :db.fn/retractAttribute :db.fn/retractEntity}]
+      (when-not (clojure.set/subset? (keys operations) known)
+        (let [unknown (apply dissoc operations known)]
+          (raise (str "Unknown operations " (apply sorted-set (keys unknown)))
```
Comment (rnewman) [inline src/datomish/db.cljc]: TIL `set` and `sorted-set` have different signatures. For no reason.

File: src/datomish/db.cljc (hunk 2 above)
Comment (ncalexan) [inline src/datomish/db.cljc]: Yes, `[op e a]`.

File: test/datomish/db_test.cljc (hunk 1 above)
Comment (ncalexan) [inline test/datomish/db_test.cljc]: Yeah, I did this mechanically.  I'll update as appropriate when I touch tests.

File: src/datomish/db.cljc (hunk 5 above)
Comment (ncalexan) [inline src/datomish/db.cljc]: INORITE  I got this wrong the first time.

**Comment by ncalexan** [comment]
Landed.

File: src/datomish/db.cljc (hunk 2 above)
Comment (rnewman) [inline src/datomish/db.cljc]: Can we change the arg name to `oeas`?

Comment (ncalexan) [inline src/datomish/db.cljc]: Sure!  Landed, but will push follow-up.

**Pull request closed without merging**